File: src/main/java/com/example/electronicdigitalsignature/utils/KeyGenerator.java
package com.example.electronicdigitalsignature.utils;

import com.example.electronicdigitalsignature.data.KeyPair;
import com.example.electronicdigitalsignature.data.PrivateKey;
import com.example.electronicdigitalsignature.data.PublicKey;

import java.math.BigInteger;
import java.security.SecureRandom;


/**
 * Класс генератор пары ключей
 */
public class KeyGenerator {
    /**
     * Битовая длина q
     */
    private static final int Q_BIT_LENGTH = 1024;
    /**
     * Уверенность в том что генерируемое число простое.
     * Вероятность того что число простое равна 1 - (1/2)^certainty
     */
    private static final int CERTAINTY = 100;
    private static final BigInteger BIG_INTEGER_TWO = new BigInteger("2");
    private static final SecureRandom RANDOM = new SecureRandom();

    /**
     * Метод для генерации пары ключей
     *
     * @return возвращает объект пары ключей
     */
    public static KeyPair generateKeyPair() {
        BigInteger q = generateQ();
        BigInteger p = generateP(q);
        BigInteger g = generateG(p, q);
        BigInteger w = new BigInteger(Q_BIT_LENGTH, RANDOM);
        BigInteger y = g.modPow(w, p);
        PublicKey publicKey = new PublicKey(p, q, g, y);
        PrivateKey privateKey = new PrivateKey(w);
        return new KeyPair(publicKey, privateKey);
    }

    /**
     * Метод генерации числа g
     */
    private static BigInteger generateG(BigInteger p, BigInteger q) {
        BigInteger g;
        BigInteger a;
        do {
            a = (BIG_INTEGER_TWO.add(new BigInteger(Q_BIT_LENGTH, CERTAINTY, RANDOM))).mod(p);
            BigInteger d = (p.subtract(BigInteger.ONE)).divide(q);
            g = a.modPow(d, p);
        } while (g.equals(BigInteger.ONE));
        return g;
    }

    /**
     * Метод генерации числа q
     *
     * @return q
     */
    private static BigInteger generateQ() {
        return new BigInteger(Q_BIT_LENGTH, CERTAINTY, RANDOM);
    }

    /**
     * Метод генерации числа p
     */
    private static BigInteger generateP(BigInteger q) {
        BigInteger qp = BigInteger.ONE;
        BigInteger p;
        do {
            p = q.multiply(qp).multiply(BIG_INTEGER_TWO).add(BigInteger.ONE);
            if (p.isProbablePrime(CERTAINTY)) break;
            qp = qp.add(BigInteger.ONE);
        } while (true);
        return p;
    }
}

File: src/main/kotlin/xyz/le30r/utils/KeyGenerator.kt
package xyz.le30r.utils

import xyz.le30r.data.KeyPair
import xyz.le30r.data.PrivateKey
import xyz.le30r.data.PublicKey
import java.math.BigInteger
import java.security.SecureRandom

/**
 * Класс генератор пары ключей
 */
object KeyGenerator {
    /**
     * Битовая длина q
     */
    private const val Q_BIT_LENGTH = 1024

    /**
     * Уверенность в том что генерируемое число простое.
     * Вероятность того что число простое равна 1 - (1/2)^certainty
     */
    private const val CERTAINTY = 100
    private val BIG_INTEGER_TWO = BigInteger("2")
    private val RANDOM = SecureRandom()

    /**
     * Метод для генерации пары ключей
     *
     * @return возвращает объект пары ключей
     */
    @JvmStatic
    fun generateKeyPair(): KeyPair {
        val q = generateQ()
        val p = generateP(q)
        val g = generateG(p, q)
        val w = BigInteger(Q_BIT_LENGTH, RANDOM)
        val y = g.modPow(w, p)
        val publicKey = PublicKey(p, q, g, y)
        val privateKey = PrivateKey(w)
        return KeyPair(publicKey, privateKey)
    }

    /**
     * Метод генерации числа g
     */
    private fun generateG(p: BigInteger, q: BigInteger): BigInteger {
        var g: BigInteger
        var a: BigInteger
        do {
            a = BIG_INTEGER_TWO.add(BigInteger(Q_BIT_LENGTH, CERTAINTY, RANDOM))
                .mod(p)
            val d = p.subtract(BigInteger.ONE).divide(q)
            g = a.modPow(d, p)
        } while (g == BigInteger.ONE)
        return g
    }

    /**
     * Метод генерации числа q
     *
     * @return q
     */
    private fun generateQ(): BigInteger {
        return BigInteger(Q_BIT_LENGTH, CERTAINTY, RANDOM)
    }

    /**
     * Метод генерации числа p
     */
    private fun generateP(q: BigInteger): BigInteger {
        var qp = BigInteger.ONE
        var p: BigInteger
        do {
            p = q.multiply(qp).multiply(BIG_INTEGER_TWO).add(BigInteger.ONE)
            if (p.isProbablePrime(CERTAINTY)) break
            qp = qp.add(BigInteger.ONE)
        } while (true)
        return p
    }
}